csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,319 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <abstract_tensor.h>
+#include <abstract_tensor_schedule.h>
+#include <ir/internal_base_nodes.h>
+#include <iter_visitor.h>
+#include <val_graph.h>
+
+namespace nvfuser {
+
+namespace {
+
+class AbstractTensorSchedule {
+ public:
+  static void apply(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph) {
+    AbstractTensorSchedule ats(abstract, concrete, graph);
+    ats.run();
+  }
+
+ private:
+  AbstractTensorSchedule(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph)
+      : abstract_(abstract), concrete_(concrete), graph_(graph) {}
+
+  void run() {
+    findNearestProducers();
+
+    // Now, for each ValGroup in abstract, find the closest producer ValGroups
+    // with entries in concrete_ids_ and replay the path from them. If none
+    // exists, then do not include this dimension in the output
+    std::vector<IterDomain*> loop_domain;
+    for (const AbstractId& abs_id : abstract_.domain) {
+      IterDomain* new_id = replayAbstractId(abs_id);
+      if (new_id == nullptr) {
+        continue;
+      }
+      loop_domain.push_back(new_id);
+    }
+    concrete_->setLoopDomain(loop_domain);
+  }
+
+  ValGroup abstractIdToValGroup(const AbstractId& abs_id) {
+    NVF_ERROR(
+        abs_id.is<ValGroupAndItsGraph>(),
+        "AbstractId must be a ValGroupAndItsGraph");
+    const ValGroupAndItsGraph& vgg = abs_id.as<ValGroupAndItsGraph>();
+    if (graph_ == nullptr) {
+      graph_ = vgg.graph;
+    } else {
+      NVF_ERROR(graph_ == vgg.graph);
+    }
+    return vgg.group;
+  }
+
+  //! Work backward from each loop IterDomain in concrete. When we find an
+  //! IterDomain* with a ValGroup in scheduled_val_groups, we map to it
+  //! (including siblings) in concrete_ids_. Assert that we do find a
+  //! scheduled ValGroup, otherwise we would not know where to place this
+  //! IterDomain.
+  //!
+  //! For example, suppose we have the following:
+  //!
+  //!   concrete_:
+  //!     root: iS0{i0} iS1{i1}
+  //!     loop: iS3{ceilDiv(i0 * i1, 16)} iS4{16}
+  //!     transforms:
+  //!       iS2 = merge(iS0, iS1)
+  //!       iS3, iS4 = split(iS2, 16)
+  //!   graph:
+  //!     ValGroup 0: 0 8
+  //!     ValGroup 1: 1 9
+  //!     ValGroup 2: 2 10
+  //!     ValGroup 3: 3
+  //!     ValGroup 4: 4
+  //!     ValGroup 5: 11
+  //!     ValGroup 6: 12
```

Comment:
What do `11` and `12` refer to?

Reply:
I guess this example has a diverging transform with `iS2` as the common input?

Reply:
I'll make this more clear. It was shorthand for `iS11` and `iS12`, similar to how it looks when printing a `ValGraph`.

This example was needlessly confusing since I made iS0 and iS5 mapped with one another, so it's a diverging transform as you say. I changed it to no longer have a diverging transform: the abstract tensor now just represents a merge and split on two axes followed by a merge of two other axes, and I show a concrete TV with three axes so that the last transform is uncomputable.

Reply:
> a concrete TV with three axes

Are you referring to `tv_`? Both root and loop domains seem to be 2D. Which three axes are you referring to?

Reply:
Ah, I meant this reply for the other code comment. In this case you're right that we have a loop transform in `tv_` that differs from what's present in `abstract_tensor_`. As discussed in another comment, I'm going to start asserting that this is not the case, and I will update this comment accordingly. Essentially instead of `findNearestProducers()` we will have something like `checkLoopIsScheduled` that checks that all the loop ValGroups of `tv_` are in the history of `abstract_tensor_`.

Reply:
OK, sounds good. Let me know when it's again ready for review.